src/render/render_texture.js

```diff
@@ -0,0 +1,48 @@
+// @flow
+
+const Buffer = require('../data/buffer');
+const VertexArrayObject = require('./vertex_array_object');
+const PosArray = require('../data/pos_array');
+
+import type Painter from './painter';
+
+class RenderTexture {
+    gl: WebGLRenderingContext;
+    texture: WebGLTexture;
+    buffer: Buffer;
+    vao: VertexArrayObject;
+
+    constructor(painter: Painter) {
+        const gl = this.gl = painter.gl;
+
+        const texture = this.texture = gl.createTexture();
+        gl.bindTexture(gl.TEXTURE_2D, texture);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
+        gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, painter.width, painter.height, 0, gl.RGBA, gl.UNSIGNED_BYTE, null);
+
+        gl.bindTexture(gl.TEXTURE_2D, null);
+
+        const array = new PosArray();
+        array.emplaceBack(0, 0);
+        array.emplaceBack(1, 0);
+        array.emplaceBack(0, 1);
+        array.emplaceBack(1, 1);
+        this.buffer = Buffer.fromStructArray(array, Buffer.BufferType.VERTEX);
+        this.vao = new VertexArrayObject();
+    }
+
+    attachToFramebuffer() {
+        const gl = this.gl;
+        gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, this.texture, 0);
```

Comment:
Ideally, we'd be using one *framebuffer* for each layer (with different color attachments, and the same depth attachment), rather than detaching/reattaching textures to the framebuffer. The reason for this is that it's expensive to "validate" that a framebuffer has correctly bound attachments. It looks like we don't do that yet in JS; in Native, we're [running `glCheckFramebufferStatus`](https://github.com/mapbox/mapbox-gl-native/blob/e8bfcd04740bbc35a1e25038fe7b7d2577792b66/src/mbgl/gl/context.cpp#L365-L392). Therefore, the API in Native is modeled so that you can't change the attachments of a framebuffer after creation.

http://www.songho.ca/opengl/gl_fbo.html claims that the *opposite* of what I said is true:

> Framebuffer object (FBO) provides an efficient switching mechanism; detach the previous framebuffer-attachable image from a FBO, and attach a new framebuffer-attachable image to the FBO. Switching framebuffer-attachable images is much faster than switching between FBOs.

However, further research suggests that this is outdated information, and no longer true for modern GPUs:

* [Intel built-in GPUs](https://software.intel.com/en-us/articles/opengl-performance-tips-swap-frame-buffer-objects-fbo-instead-of-surfaces-using-a-single)

  > Switching an entire FBO is more efficient than switching individual surfaces one at a time.

* [PowerVR](http://cdn.imgtec.com/sdk-documentation/PowerVR.Performance+Recommendations.pdf)

  > For optimal performance, [...] attachments should not be added or removed once the FBO has been created.

* [nVidia](https://developer.nvidia.com/sites/default/files/akamai/gamedev/docs/Porting%20Source%20to%20Linux.pdf)

  > Do not create a single FBO and then swap out attachments on it. This causes lots of validation in the driver, which in turn leads to poor performance.

[ARM/Mali](https://community.arm.com/graphics/b/blog/posts/mali-performance-2-how-to-correctly-handle-framebuffers) also has a very good article on framebuffer handling and how it interacts with buffer swapping that also confirms the behavior we've seen of PowerVR GPUs which prompted https://github.com/mapbox/mapbox-gl-native/issues/9286

Reply:
@kkaefer I implemented this in https://github.com/mapbox/mapbox-gl-js/pull/5101/commits/0651dcfa32fec3bd4d453f5cbf8008ff25ddf1ac. We still have to detach/reattach the depth renderbuffer between framebuffers, so I do wonder about performance gains — before there was one framebuffer with an always-attached renderbuffer and we switched texture attachments, and now there are many framebuffers with their own texture but we switch renderbuffer attachments, so I assume they still do validation every time we attach the depth rbo… 🤔

Reply:
Hmm, why are we attaching the depth renderbuffer for every render? Can we leave the same renderbuffer attached to multiple framebuffers?

Reply:
Oh, interesting — okay, I guess I had conceptually assumed that since `gl.framebufferRenderbuffer` attaches a renderbuffer to a framebuffer, that attachment was unique to the bound framebuffer. It seems instead based on what you're implying (and manual testing confirms this) that that attaches a renderbuffer to whatever is bound at the `gl.FRAMEBUFFER` attachment point at the time of the render?
To me [this](https://www.khronos.org/registry/OpenGL-Refpages/es2.0/xhtml/glFramebufferRenderbuffer.xml) doesn't read super clearly either way —
> attach a renderbuffer object to a framebuffer object

> glFramebufferRenderbuffer attaches the renderbuffer specified by renderbuffer as one of the logical buffers of the currently bound framebuffer object

but it does seem to work when I only attach the depth RBO on the first fill-extrusion draw…

Reply:
hmm, what could work is upon creation of a framebuffer, attach both the depth RBO + color texture to the FBO, with every FBO having a different texture, but the same depth RBO?

Reply:
Right — this is how it currently works